homeassistant/components/tuya/alarm_control_panel.py

```diff
@@ -112,9 +118,23 @@ def __init__(
     @property
     def alarm_state(self) -> AlarmControlPanelState | None:
         """Return the state of the device."""
-        if not (status := self.device.status.get(self.entity_description.key)):
-            return None
-        return STATE_MAPPING.get(status)
+        mode = self.device.status.get(DPCode.MASTER_MODE)
+        state = self.device.status.get(DPCode.MASTER_STATE)
+        # When the alarm is triggered, only its 'state' is changing. From 'normal' to 'alarm'.
+        # The 'mode' doesn't change, and stays as 'arm' or 'home'.
+        if state == State.ALARM:
+            return AlarmControlPanelState.TRIGGERED
+        return STATE_MAPPING.get(mode)
+
+    @property
+    def changed_by(self) -> str | None:
+        """Last change triggered by."""
+        state = self.device.status.get(DPCode.MASTER_STATE)
```

Comment:
Same as above, let's find the DP in the init, so we can just reference it here directly.

Reply:
Removed all hard-coded DP's and added a search in the init function.